kore/src/Kore/Step/Axiom/Matcher.hs

```diff
@@ -435,3 +452,160 @@ checkVariableEscape vars predSubst
+        :: Int
+        -> Builtin.InternalList (TermLike variable)
+        ->  ( Builtin.InternalList (TermLike variable)
+            , Builtin.InternalList (TermLike variable)
+            )
+    splitByL2 idx l =
+        Bifunctor.bimap
+            (updateInnerList l)
+            (updateInnerList l)
+            $ Seq.splitAt idx (Builtin.builtinListChild l)
+
+    updateInnerList l newValue = l { Builtin.builtinListChild = newValue }
+matchAppBuiltins qv symbol args (Builtin.BuiltinMap m2)
+  | symbol == Builtin.builtinMapConcat m2 =
+    case args of
+        [ Builtin_ (Builtin.BuiltinMap m1), x@(Var_ _) ] -> matchMaps m1 x
+        [ x@(Var_ _), Builtin_ (Builtin.BuiltinMap m1) ] -> matchMaps m1 x
+        _ -> nothing
+  where
+    matchMaps
+        :: Builtin.InternalMap (TermLike Concrete) (TermLike variable)
+        -> TermLike variable
+        -> MaybeT unifier (Predicate variable)
+    matchMaps m1 x = do
+        let (prefix2, suffix2) = splitM2By m1
+        prefix  <-
+            matchBuiltins
+                qv
+                (Builtin.BuiltinMap m1)
+                (Builtin.BuiltinMap prefix2)
+        suffix <- match qv x (mkBuiltin $ Builtin.BuiltinMap suffix2)
+        pure $ prefix <> suffix
+
+    splitM2By
+        :: Builtin.InternalMap (TermLike Concrete) (TermLike variable)
+        ->  ( Builtin.InternalMap (TermLike Concrete) (TermLike variable)
+            , Builtin.InternalMap (TermLike Concrete) (TermLike variable)
+            )
+    splitM2By m =
+        Bifunctor.bimap
+            (updateInnerMap m)
+            (updateInnerMap m)
+            $ Map.partitionWithKey (memberOfMap m) (Builtin.builtinMapChild m2)
+
+    updateInnerMap m newValue = m { Builtin.builtinMapChild = newValue }
+    memberOfMap m key _value = Map.member key $ Builtin.builtinMapChild m
+matchAppBuiltins _ _ _ _ = nothing
+
+matchBuiltins
+    :: FreshVariable variable
+    => Show variable
+    => Unparse variable
+    => SortedVariable variable
+    => MonadUnify unifier
+    => Map.Map variable variable
+    -> Builtin.Builtin (TermLike Concrete) (TermLike variable)
+    -> Builtin.Builtin (TermLike Concrete) (TermLike variable)
+    -> MaybeT unifier (Predicate variable)
+matchBuiltins qv (Builtin.BuiltinList l1) (Builtin.BuiltinList l2)
+  | ((==) `on` (Seq.length)) seq1 seq2 =
+    fmap Foldable.fold
+        . traverse match'
+        $ Seq.zip seq1 seq2
+  where
+    seq1 = Builtin.builtinListChild l1
+    seq2 = Builtin.builtinListChild l2
+    match' = uncurry (match qv)
+matchBuiltins qv (Builtin.BuiltinSet s1) (Builtin.BuiltinSet s2)
+  | varElems2 == [] && varSets2 == [] && length varSets1 <= 1 = do
+    let remainder = Set.toList $ concrete2 Set.\\ concrete1
+    Monad.guard $ length remainder + length concrete1 == length concrete2
+    Monad.guard $ length remainder >= length varElems1
+
+    variableMatches <-
+        fmap Foldable.fold
+            . traverse match'
+            . zip varElems1
```

Comment:
`zip` discards the tail of the longer list; is that OK here? It isn't clear to me.

Reply:
Yes, I think it's precisely what we need:
- if we have more `varElems`, code is not reached due to the `guard` above
- if we have more concrete elements than `varElems`, that is fine, since we check the `setRemainder` below